crates/storage/codecs/derive/src/arbitrary.rs

```diff
@@ -0,0 +1,63 @@
+use proc_macro::{self, TokenStream};
+use proc_macro2::TokenStream as TokenStream2;
+use quote::{format_ident, quote};
+use syn::DeriveInput;
+
+/// If `compact` or `rlp` is passed to `derive_arbitrary`, this function will generate the
+/// corresponding proptest roundtrip tests.
+pub fn maybe_generate_tests(args: TokenStream, ast: &DeriveInput) -> TokenStream2 {
+    let type_ident = ast.ident.clone();
+
+    let mut traits = vec![];
+    let mut roundtrips = vec![];
+
+    for arg in args {
+        if arg.to_string() == "compact" {
+            traits.push(quote! { use super::Compact; });
+            roundtrips.push(quote! {
+                {
+                    let mut buf = vec![];
+
+                    let len = field.clone().to_compact(&mut buf);
+                    let (decoded, _) = super::#type_ident::from_compact(&buf, len);
+
+                    assert!(field == decoded);
+                }
+            });
+        } else if arg.to_string() == "rlp" {
+            traits.push(quote! { use reth_rlp::{Encodable, Decodable}; });
+            roundtrips.push(quote! {
+                {
+                    let mut buf = vec![];
+
+                    let len = field.clone().encode(&mut buf);
+                    let decoded = super::#type_ident::decode(&mut buf.as_slice()).unwrap();
+
+                    assert!(field == decoded);
+                }
+            });
+        }
+    }
+
+    let mut tests = TokenStream2::default();
+    if !roundtrips.is_empty() {
+        let mod_tests = format_ident!("{}Tests", ast.ident);
+
+        tests = quote! {
+            #[allow(non_snake_case)]
+            #[cfg(test)]
+            mod #mod_tests {
+                #(#traits)*
+
+                #[test]
+                fn proptest() {
+                    proptest::proptest!(proptest::prelude::ProptestConfig::with_cases(1000), |(field: super::#type_ident)| {
```

Comment:
What would be the default number of cases?

Fields with `Vec<Vec<_>>` usually take quite long... Would an extra `arg` to define custom number of cases be useful for these types?

Reply:
could we add this as an arg to the proc-macro?

Reply:
I don't mind doing in follow up

Reply:
we have to do now because it really blows up. Some types will have many vec structs of vec structs...